Add function 'groupByNodes' to group on several indexes

gboily opened the pull request:
Add a new function 'groupByNodes' that support multiple indexes like the function 'aliasByNode'.

This avoids to add a new graph target for each new metric/node created.

Ex:
&target=groupByNodes(ganglia.**server`*`**.`*`.cpu.**load`*`**,"sumSeries",1,4)

Will "generate" the following formula : 
- sumSeries(ganglia.server1.*.cpu.load5)
- sumSeries(ganglia.server1.*.cpu.load10)
- sumSeries(ganglia.server1.*.cpu.load15)
- sumSeries(ganglia.server2.*.cpu.load5)
- sumSeries(ganglia.server2.*.cpu.load10)
- sumSeries(ganglia.server2.*.cpu.load15)
- ...

Will produce the following legends : 
- server1.load5
- server1.load10
- server1.load15
- ...
- server2.load5
- server2.load10
- server2.load15
- ...

steve-dave commented:
I can see how this would be useful, but could it be achieved by adding the functionality to groupByNode() and maybe aliasing groupByNodes() to groupByNode()?

gboily commented:
Yes, it would be possible to merge some code, but due to the callback parameter being not at the same place, we must either : 
- keep the function groupByNodes and modify the function groupByNode to call groupByNodes
```
def groupByNode(requestContext, seriesList, nodeNum, callback)
  groupByNodes(requestContext, seriesList, callback, nodeNum)

def groupByNodes(requestContext, seriesList, callback, *nodes)
  ...
```
- or we can keep only the groupByNode function and accept parameters in several orders by detecting their types, etc (a bit ugly IMO)
```
def groupByNode(requestContext, seriesList, *args)
  if isinstance(args[0], int)
    ...
```

Any preference/other  idea ?

steve-dave commented:
Possibly use `*args` and require that the callback always be the last parameter or alternatively do it the way you have proposed, but strip the code from `groupByNode()` and convert it to a wrapper of `groupByNodes()` so that it's backwards compatible but we avoid code duplication.

What do you think @obfuscurity ?

obfuscurity commented:
@steve-dave Yeah, :+1: on your proposed change. That seems like the "least ugly" of all choices. But we need some very solid test coverage for this merged function to cover both use cases.

steve-dave commented:
@gboily can you please convert `groupByNode()` to a wrapper of `groupByNodes()` and ensure there are good tests?

gboily commented:
@steve-dave I've updated groupByNode and added tests.
I was not able to use "sumSeries" (nor other *Series I guess) so I used "keepLastValue" instead.
So, tests only test aliases computation and not really calculation results as keepLastValue does not change the values in that case.

Here is the error when using "sumSeries" : 
```
ERROR: test_groupByNodes (tests.test_functions.FunctionsTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/src/graphite-web2/webapp/tests/test_functions.py", line 291, in test_groupByNodes
    verify_groupByNodes(expectedResult, 1)
  File "/usr/local/src/graphite-web2/webapp/tests/test_functions.py", line 281, in verify_groupByNodes
    results = functions.groupByNodes({}, copy.deepcopy(seriesList), "sumSeries", *nodes)
  File "/usr/local/src/graphite-web2/webapp/graphite/render/functions.py", line 2855, in groupByNodes
    metaSeries[key])[0]
  File "/usr/local/src/graphite-web2/webapp/graphite/render/functions.py", line 191, in sumSeries
    name = "sumSeries(%s)" % formatPathExpressions(seriesList)
  File "/usr/local/src/graphite-web2/webapp/graphite/render/functions.py", line 157, in formatPathExpressions
    [pathExpressions.append(s.pathExpression) for s in seriesList if not pathExpressions.count(s.pathExpression)]
AttributeError: 'TimeSeries' object has no attribute 'pathExpression'
```

I also got a timezone problem and temporaly disabled test_render.test.
It may be due to running tests in a docker container (I tried changing my timezone to GMT and GMT+1 without success) :  
```
# tox -e py27-django14 -- /usr/local/src/graphite-web2/webapp/tests/test_functions.py:FunctionsTest
GLOB sdist-make: /usr/local/src/graphite-web2/setup.py
py27-django14 inst-nodeps: /usr/local/src/graphite-web2/.tox/dist/graphite-web-0.10.0-alpha.zip
py27-django14 runtests: PYTHONHASHSEED='1309592725'
py27-django14 runtests: commands[0] | django-admin.py test
/usr/local/src/graphite-web2/webapp/graphite/settings.py:233: UserWarning: SECRET_KEY is set to an unsafe default. This should be set in local_settings.py for better security
  warn('SECRET_KEY is set to an unsafe default. This should be set in local_settings.py for better security')
Creating test database for alias 'default'...
....................................F.../usr/local/src/graphite-web2/.tox/py27-django14/local/lib/python2.7/site-packages/django/db/models/fields/__init__.py:808: RuntimeWarning: DateTimeField received a naive datetime (1970-01-01 01:00:00) while time zone support is active.
  RuntimeWarning)
.............
======================================================================
FAIL: test_correct_timezone (tests.test_render.RenderTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/src/graphite-web2/webapp/tests/test_render.py", line 122, in test_correct_timezone
    self.assertEqual(data, expected)
AssertionError: Lists differ: [[12, 1393394460], [12, 139339... != [[12, 1393398060], [12, 139340...

First differing element 0:
[12, 1393394460]
[12, 1393398060]

- [[12, 1393394460], [12, 1393398060]]
+ [[12, 1393398060], [12, 1393401660]]

----------------------------------------------------------------------
Ran 53 tests in 0.251s

FAILED (failures=1)
```

robertclancy commented:
This would be very useful. What needs to happen to make this mergeable?

obfuscurity commented:
@robertclancy We'd need to first rebase it, then resolve the issues reported by @gboily and put it through some actually test use.

Dieterbe commented on webapp/graphite/render/functions.py:
```diff
@@ -2821,10 +2821,30 @@ def groupByNode(requestContext, seriesList, nodeNum, callback):
+
+def groupByNodes(requestContext, seriesList, callback, *nodes):
+  """
+  Takes a serieslist and maps a callback to subgroups within as defined by a common node
```
or multiple nodes

Dieterbe commented on webapp/graphite/render/functions.py:
```diff
@@ -2821,10 +2821,30 @@ def groupByNode(requestContext, seriesList, nodeNum, callback):
+    nodes=[nodes]
   for series in seriesList:
-    key = series.name.split(".")[nodeNum]
+    key = '.'.join(series.name.split(".")[n] for n in nodes)
```
I'm a bit surprised / confused by how this works. I understand `'.'.join(<somelist>` and i've done list comprehensions before, but it's interesting how the list here is constructed with the `<list>[n]` syntax wherein `[n]` is part of a list comprehension.

```
>>> name = "abc.def.ghi.jkl"
>>> nodes = [0,2]
>>> name.split(".")[n] for n in nodes
  File "<stdin>", line 1
    name.split(".")[n] for n in nodes
                         ^
SyntaxError: invalid syntax
>>> ".".join(name.split(".")[n] for n in nodes)
'abc.ghi'
>>> 
```

anyway, just some python magic / syntactic sugar i guess. it seems to work fine.

Dieterbe commented:
looks good to me. i need this. i applied the patch to functions.py and am already using it in prod :+1:

bmhatfield commented:
@gboily Thanks for your PR! We've had a few similar PRs and seem to have settled on yours to solve the problem. Would you mind dropping by this thread and rebasing so we can look towards merging this?

Thank you!

bmhatfield commented:
@gboily One last ping on this; otherwise we'll probably take over this PR / patch and make it mergable.

@Dieterbe perhaps you'd be best positioned to turn this into a mergable PR given that you've already done it against your system?

ctrochalakis commented:
I am also willing to help if @gboily and @Dieterbe are busy.

bmhatfield commented:
@ctrochalakis go for it :-)
